solutions/CSES Planets Cycles.mdx

```diff
@@ -0,0 +1,209 @@
+
+<LanguageSection>
+
+<CPPSection>
+
+## C++ Implementation
+
+```cpp
+#include <iostream>
+#include <queue>
+using namespace std;
+void dfs(int planet);
+
+bool visited[200000]{};
+int destinations[200000];
+int pathlength[200000]{};
+queue<int> path;
+int steps = 0;
+
+int main()
+{
+	int n;
+	cin >> n;
+	for (int i = 0; i < n; i++)
+	{
+		cin >> destinations[i];
+		destinations[i]--;
+	}
+	for (int i = 0; i < n; i++)
+	{
+		if (!visited[i])
+		{
+			steps = 0;
+			dfs(i);
+			int decrement = 1;
+			//for each planet in current path, calculate pathlength
+			while (!path.empty())
+			{
+				//we are in the cycle; all nodes have same pathlength
+				if (path.front() == path.back())
+				{
+					decrement = 0;
+				}
+				pathlength[path.front()] = steps;
+				steps -= decrement;
+				path.pop();
+			}
+		}
+	}
+	for (int i = 0; i < n; i++)
+	{
+		cout << pathlength[i] << " ";
+	}
+	cout << endl;
+	return 0;
+}
+
+void dfs(int planet)
+{
+	//add planet to path
+	path.push(planet);
+	if (visited[planet])
+	{
+		//add pathlength of the repeat planet to current step count
+		steps += pathlength[planet];
+		return;
+	}
+	visited[planet] = true;
+	steps++;
+	dfs(destinations[planet]);
+}
+```
+
+</CPPSection>
+
+<JavaSection>
+
+## Java Implementation
+
+```java
+
+import java.util.*;
+import java.io.*;
+
+public class planetCycles {
+    static int N, steps;
+    static int[] destinations, pathlength;
+    static boolean[] visited;
+    static LinkedList<Integer> path = new LinkedList<>();
+    public static void main(String[] args) throws IOException {
+        Kattio in = new Kattio(System.in);
+
+        N = in.getInt();
+        destinations = new int[N+1];
+        visited = new boolean[N+1];
+        pathlength = new int[N+1];
+        for (int i = 1; i < N + 1; i++) {
+            destinations[i] = in.getInt();
+        }
+
+        for (int i = 1; i < N + 1; i++) {
+            if(!visited[i]){
+                steps = 0;
+                dfs(i);
+                int decrement = 1;
+                Integer last = path.peekLast();
+                while(!path.isEmpty()){
+                    if(path.peekFirst().equals(last)) decrement = 0;
+                    pathlength[path.poll()] = steps;
+                    steps-=decrement;
+                }
+            }
+        }
+
+        for (int i = 1; i < N + 1; i++) {
+            System.out.print(pathlength[i] + " ");
+        }
+    }
+
+    public static void dfs(int n){
+        visited[n] = true;
+        path.add(n);
+        steps++;
+        if(!visited[destinations[n]])
+            dfs(destinations[n]);
+        else
+            path.add(destinations[n]); steps += pathlength[destinations[n]];
+
+    }
+
+    static class Kattio extends PrintWriter {
```

Comment:
would prefer the current version of `Kattio` (https://usaco.guide/general/io?lang=java#java). any particular reason why this is contained within the public class (as opposed to preceding it?)